archive_converter.py

```diff
@@ -14,65 +14,71 @@
 
 import email
 import os
-from typing import List, Dict
+import re
+from typing import List, Dict, Tuple, Set
 from setup_gmail import Message
+from message_dao import MessageDao
 
 class ArchiveMessageIndex(object):
-    def __init__(self):
-        # Maps message subject (str) to starting message of email thread
-        self._key_to_thread_start : Dict[str, Message] = {}
-        # Maps message.id to message
-        self._messages_seen : Dict[str, Message] = {}
-
-    def find_thread(self, key: str) -> Message:
-        return self._key_to_thread_start[key]
-
-    def update(self, data_dir: str):
+    def __init__(self, message_dao : MessageDao):
+        self._message_dao = message_dao
+        
+    def update(self, data_dir: str) -> Tuple[Tuple[Message, ...], Tuple[Message, ...]]:
         new_messages : List[Message] = []
 
         for filename in os.listdir(data_dir):
             if not filename.endswith(".txt"):
                 continue
             email = generate_email_from_file(os.path.join(data_dir, filename))
-            if email.id not in self._messages_seen:
+            if not self._message_dao.get(email.id):
                 new_messages.append(email)
-        self._add_messages_to_index(new_messages)
+        return self._add_messages_to_index(new_messages)
```

Comment:
Is there a reason we can't just return `new_messages` here?
Do we need to distinguish between the two different types of new messages?

If so, let's consider augmenting the commit message with a description and/or a code comment.

Reply:
so I think it would be fine to return new_messages, but that just leaves the processing into two different types of new_messages for later? I think we need two different types of new messages because not all new messages are directly uploaded. My current understanding (which may be wrong) is that comments are added as children to parent patches and then the upload_all_comments function is called. Also, patchsets are all uploaded together so some processing needs to be done to see whether the start of the patchset is also within the current batch of emails. If it is, then only the start of the patchset needs to be uploaded (as it will upload the rest as well). If it is not, then the reply must be uploaded individually because the start of the patchset has already been uploaded and should not be uploaded twice.

Reply:
Meta: adding some '\n' in comments makes it a bit easier to read :)

Yes, we need to treat patches different from comments.
We'd want to upload all patches first and then add comments in the right places.

But we can distinguish between the types of messages later.
As suggested below, the check you're doing now could be moved into the message class.

We could then partition the messages into the two types afterwards fairly easily with that helper func.
So this function doesn't seem like it needs to take on this additional responsibility.

General tip: The more things a function/class is responsible, the harder it is to understand.

Reply:
so for now, would you say it's best to just write a todo and have it return new_messages?

Reply:
SGTM.